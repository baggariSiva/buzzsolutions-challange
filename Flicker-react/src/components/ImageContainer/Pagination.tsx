import React from "react";
import TablePagination from "@mui/material/TablePagination";

type Props = {
  page: number;
  pageSize: number;
  setPage: (page: number) => void;
  setPageSize: (pageSize: number) => void;
  totalRecordCount: number;
};

export default function Pagination({
  page,
  pageSize,
  setPage,
  setPageSize,
  totalRecordCount,
}: Props): JSX.Element {
  const handleChangePage = (
    _: React.MouseEvent<HTMLButtonElement> | null,
    newPage: number
  ) => {
    setPage(newPage);
  };

  const handleChangeRowsPerPage = (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    setPageSize(parseInt(event.target.value, 10));
    setPage(0);
  };

  return (
    <TablePagination
      component="div"
      count={totalRecordCount}
      page={page}
      onPageChange={handleChangePage}
      rowsPerPage={pageSize}
      onRowsPerPageChange={handleChangeRowsPerPage}
    />
  );
}
