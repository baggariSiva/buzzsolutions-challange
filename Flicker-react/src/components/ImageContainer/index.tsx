import { useEffect, useMemo, useState } from "react";

import Pagination from "./Pagination";
import type { ImageData, Response } from "../../types";

type Props = {
  searchQuery: string;
};

const API_URL: string = `http://localhost:5000/api/v1/photos`;

const INITIAL_DATA: Response = {
  photos: {
    page: 1,
    pages: 1,
    perpage: 10,
    photo: [],
    total: 10,
  },
};

function ImageContainer({ searchQuery }: Props) {
  const [page, setPage] = useState<number>(0);
  const [pageSize, setPageSize] = useState<number>(10);
  const [imageData, setImageData] = useState<Response>(INITIAL_DATA);

  const params = useMemo((): string => {
    const search = searchQuery ? `search=${searchQuery}` : null;
    const pageNo = `page=${page}`;
    const perPage = `perPage=${pageSize}`;

    return [search, pageNo, perPage].filter((param) => param).join("&");
  }, [page, pageSize, searchQuery]);

  useEffect(() => {
    const fetchData = async () => {
      try{
        const response = await fetch(`${API_URL}?${params}`);
        const data = await response.json();
      setImageData(data);
      }catch(error){
        console.log("Error in fetching data",error)
      }
     
    };

    fetchData();
  }, [params]);

  return (
    <div className="container text-center">
      <div className="row row-cols-2 row-cols-lg-5 g-2 g-lg-3">
        {imageData?.photos?.photo?.map((image: ImageData, index: number) => {
          return (
            <div className="col" key={index}>
              <img
                src={image.url_c}
                className="img-thumbnail"
                alt={image.title}
                object-fit="contain"
              />
            </div>
          );
        })}
        {imageData?.photos?.photo?.length === 0 && !!searchQuery && (
          <div className="no-img">
            <p>No images found</p>
          </div>
        )}
      </div>
      {imageData?.photos?.photo.length > 0 && (
        <Pagination
          page={page}
          pageSize={pageSize}
          setPage={setPage}
          setPageSize={setPageSize}
          totalRecordCount={imageData.photos.total}
        />
      )}
    </div>
  );
}

export default ImageContainer;
