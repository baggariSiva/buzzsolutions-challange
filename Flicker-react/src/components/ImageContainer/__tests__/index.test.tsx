import { render, screen } from "@testing-library/react";
import ImageContainer from "../";

describe("ImageContainer", () => {
  it("renders without crashing", () => {
    render(<ImageContainer searchQuery="test" />);
    expect(screen.getByText(/No images found/i)).toBeInTheDocument();
  });

    it("matches snapshot", () => {
        const { asFragment } = render(<ImageContainer searchQuery="test" />);
        expect(asFragment()).toMatchSnapshot();
    });

    it("displays images when api data is fetched", async () => {
        const mockData = {
            photos: {
                page: 1,
                pages: 1,
                perpage: 10,
                photo: [
                    {
                        id: "1",
                        title: "test1",
                        url_c: "http://test1.com",
                    },
                    {
                        id: "2",
                        title: "test2",
                        url_c: "http://test2.com",
                    },
                ],
                total: 2,
            },
        };
        jest.spyOn(global, "fetch").mockResolvedValue({
            json: jest.fn().mockResolvedValue(mockData),
            ok: true,
        } as any);

        render(<ImageContainer searchQuery="test" />);
        expect(await screen.findByAltText("test1")).toBeInTheDocument();
        expect(await screen.findByAltText("test2")).toBeInTheDocument();
    });

    it("displays no images found when api data is empty", async () => {
        const mockData = {
            photos: {
                page: 1,
                pages: 1,
                perpage: 10,
                photo: [],
                total: 0,
            },
        };
        jest.spyOn(global, "fetch").mockResolvedValue({
            json: jest.fn().mockResolvedValue(mockData),
            ok: true,
        } as any);

        render(<ImageContainer searchQuery="test" />);
        expect(await screen.findByText(/No images found/i)).toBeInTheDocument();
    });
});
