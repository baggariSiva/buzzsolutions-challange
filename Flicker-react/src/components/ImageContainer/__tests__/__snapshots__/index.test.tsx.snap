// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageContainer matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container text-center"
  >
    <div
      class="row row-cols-2 row-cols-lg-5 g-2 g-lg-3"
    >
      <div
        class="no-img"
      >
        <p>
          No images found
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
